### PR TITLE
Add widgets

### LINKED ISSUE #458: widget
from an email we got (and i think we should definitely think about this, these are some good ideas masha'Allah):

Assalamu Alaikum and jazakumumullahu katheeran

I would like to urge the team to not miss a great opportunity for much reward. Two features _in a widget_ form will do wonders for the ppl of Quran, the family of Allah, the best Muslims:
1. Direct Surah/Ayah jump: People learning islam and daa'ees especially find references to Aayaat on the move and need to cross check or quote it or bookmark for tafsir study. Readers of physical mushaf at home might want to continue at the mosque for short reading. A widget that takes input like (23:21) and then jumps or book marks or shares immediately will be super.
2. Arabic search jump: for same read on move or quote on move  as above - on the move you want to look up ayat...opening app takes time. Widget launches straight to translate/bookmark/share

So while typing this email i could quote ayaat very quickly, especially if you can design a keyboard add-on (that's third feature actually?)

Jkkk

Salam.

### PR DESCRIPTION
**Pull request by ismail-s**
Add 2 widgets, a search widget and a bookmarks widget. See #458 and #845.

Closes #458
Closes #845

Rather than starting from scratch, I built on top of the work in #845 (Jazakallah @nyitgroup for your previous work).

### CODE REVIEW

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  init {
+    (context.applicationContext as QuranApplication).applicationComponent.inject(this)
+    populateListItem()
+  }
+
+  private fun populateListItem() {
+    val appContext = context.applicationContext
+    val mBookmarksDBAdapter = BookmarksDBAdapter(appContext, quranInfo.numberOfPages)
+    val bookmarksList = mBookmarksDBAdapter.getBookmarks(BookmarksDBAdapter.SORT_DATE_ADDED)
```
Comment (ismail-s) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt:33]: I could change this to sort by location if people want, that may seem more sensible.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetService.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.quran.labs.androidquran.widgets
+
+import android.content.Intent
+import android.widget.RemoteViewsService
+
+class WidgetService : RemoteViewsService() {
+  /*
+* So pretty simple just defining the Adapter of the listview
+* here Adapter is ListProvider
+* */
```
Comment (ismail-s) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetService.kt:10]: Reminder to self to tidy this up.

File: app/src/main/res/xml/bookmarks_widget_info.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<appwidget-provider xmlns:android="http://schemas.android.com/apk/res/android"
+    android:initialKeyguardLayout="@layout/bookmarks_widget"
+    android:initialLayout="@layout/bookmarks_widget"
+    android:minHeight="110dp"
+    android:minWidth="180dp"
+    android:previewImage="@drawable/bookmarks_appwidget_preview"
```
Comment (ismail-s) [inline app/src/main/res/xml/bookmarks_widget_info.xml:7]: Reminder to self to look into updating these preview images.

File: app/src/main/res/layout/bookmarks_widget_list_row.xml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    android:layout_marginBottom="8dp"
+    android:layout_marginTop="8dp"
+    android:orientation="horizontal">
+
+  <ImageView
+      android:id="@+id/widget_favorite_icon"
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:layout_gravity="center_vertical"
+      android:layout_marginLeft="@dimen/audiobar_separator_padding"
```
Comment (ismail-s) [inline app/src/main/res/layout/bookmarks_widget_list_row.xml:17]: Reminder to self to check if there's a better margin constant to use. And to more generally go over this file as a whole.

File: app/src/main/res/layout/search.xml [hunk 5]
```diff
@@ -9,18 +9,28 @@
...
     	android:layout_width="match_parent" android:layout_height="wrap_content" android:visibility="gone"/>
-    <TextView android:id="@+id/search_area"
-            android:textColor="?android:textColorPrimary"
-            android:textSize="17sp"
-            android:text="@string/search_key"
-            android:background="@android:drawable/title_bar"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content" />
+
+  <androidx.appcompat.widget.SearchView
```
Comment (ismail-s) [inline app/src/main/res/layout/search.xml:15]: #845 had an `EditText` here, I changed it to a `SearchView`.

File: app/src/main/res/layout/search_widget.xml [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:background="@drawable/search_widget_background"
```
Comment (ismail-s) [inline app/src/main/res/layout/search_widget.xml:4]: I spent a bit of time trying to make the widget have round corners, and ended up setting this custom background. It's not really a good way of doing this though, I was tempted to switch it back to as it was in #845, but thought I'd put it in this PR instead and see what people think. I haven't rounded the corners of the bookmarks widget, with this approach it felt like it'd probably be more complicated (and I think that widget looks ok as-is).

File: app/src/main/res/values/colors.xml [hunk 7]
```diff
@@ -11,12 +11,12 @@
...
   <color name="sura_number_color">#7fffffff</color>
   <color name="sura_details_color">#b2ffffff</color>
-  <color name="transparent_actionbar_color">#AA000000</color>
+  <color name="transparent_actionbar_color">#aa000000</color>
   <color name="panel_background_color">#ff424242</color>
   <color name="translation_hdr_color">#ff777777</color>
   <color name="translation_sura_header">#fff1ecd8</color>
   <color name="translation_sura_header_night">#ff424242</color>
-  <color name="translation_verse_text_color">#ff858383</color>
+  <color name="translation_verse_text_color">#858383</color>
```
Comment (ismail-s) [inline app/src/main/res/values/colors.xml:19]: Reminder to self to check why these changes were made to this file and revert if necessary.

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -407,6 +407,9 @@
 
   <!-- format strings -->
   <string name="sura_ayah" translatable="false">%1$d:%2$d</string>
+  <string name="bookmarks_appwidget_text">Quran Bookmarks</string>
+  <string name="add_widget">Add widget</string>
+  <string name="search_appwidget_text">Quran Search </string>
```
Comment (ismail-s) [inline app/src/main/res/values/strings.xml:412]: Reminder to self to remove these strings if they're not being used.

**Comment by ahmedre** [comment]
salam 3alaikum,
just tried this - masha'Allah great so far! i think this is very promising masha'Allah, may Allah reward you greatly!

i have some feedback (haven't looked at the code yet though) -

- bookmarks - should we / can we also pick up the "ayah bookmarks?"
- if there are no bookmarks, what do you think about falling back to "recents" or some suggested items - like say sura Kahf on Fridays?
- search is not working - it's opening the search results page, but i can't get to typing something for search to happen 
- there might be a memory leak somewhere, we need to check in sha' Allah

walsalam 3alaikum.

File: app/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -163,11 +163,35 @@
         <action android:name="android.intent.action.MEDIA_BUTTON"/>
       </intent-filter>
     </service>
+    <service
+        android:name=".widgets.WidgetService"
```
[the remark this answers is not in the document]
Comment (ahmedre) [reply app/src/main/AndroidManifest.xml:167]: yeah - we can add Robolectric, but i am not sure how much you can test widgets with it - ideally i'd say let's just test any new "business logic" with just vanilla unit tests - since hopefully anything left is just the minimum required code for data / layout to the widget.

File: app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.kt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    for (appWidgetId in appWidgetIds) {
+      val serviceIntent = Intent(context, WidgetService::class.java).apply {
+        putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+        data = Uri.parse(toUri(Intent.URI_INTENT_SCHEME))
+      }
+
+      val widget = RemoteViews(context.packageName, R.layout.bookmarks_widget)
+
+      var intent = Intent(context, QuranDataActivity::class.java)
+      var pendingIntent = PendingIntent.getActivity(context, 0, intent, 0)
```
[the remark this answers is not in the document]
Comment (ahmedre) [reply app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.kt:27]: i'd vote for val over var whenever possible since it makes things cleaner generally - though as you said, in this particular case, it's not bad - but if you want to, i wouldn't mind :)

File: app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.kt [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      widget.setEmptyView(R.id.list_view_widget, R.id.empty_view)
+
+      appWidgetManager.updateAppWidget(appWidgetId, widget)
+    }
+    super.onUpdate(context, appWidgetManager, appWidgetIds)
+  }
+
+  override fun onReceive(context: Context, intent: Intent) {
+    super.onReceive(context, intent)
+    updateWidget(context)
```
[the remark this answers is not in the document]
Comment (ahmedre) [reply app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.kt:55]: according to the docs:

> This is called for every broadcast and before each of the above callback methods. You normally don't need to implement this method because the default AppWidgetProvider implementation filters all App Widget broadcasts and calls the above methods as appropriate.

from [here](https://developer.android.com/guide/topics/appwidgets)

File: app/src/main/java/com/quran/labs/androidquran/SearchActivity.java [hunk 12]
```diff
@@ -169,6 +174,9 @@ public void onLoadFinished(@NonNull Loader<Cursor> loader, Cursor cursor) {
         buttonGetTranslations.setText(R.string.get_translations);
         buttonGetTranslations.setVisibility(View.VISIBLE);
       }
+      if (adapter != null) {
+        adapter.swapCursor(null);
```
[the remark this answers is not in the document]
Comment (ahmedre) [reply app/src/main/java/com/quran/labs/androidquran/SearchActivity.java:178]: wow, great catch!
to be honest we should stop using CursorAdapter here completely :( this whole class should be rewritten at some point :(

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java [hunk 13]
```diff
@@ -144,6 +147,7 @@ public void replaceRecentRangeWithPage(int deleteRangeStart, int deleteRangeEnd,
   }
 
   public void addRecentPage(int page) {
+    BookmarksWidget.updateWidget(context);
```
[the remark this answers is not in the document]
Comment (ahmedre) [reply app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java:150]: take a look at `BookmarkPresenter` i think - there's an rx observable there. we can just subscribe to it and refresh then - not sure what the right place to subscribe is, but in worst case, some object that subscribes whenever there is at least one active widget

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    suraPageItemList = bookmarksList.filter { it.isPageBookmark() }.map {
```
[the remark this answers is not in the document]
Comment (ahmedre) [reply app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt:34]: what makes it tricky in this case to handle both (even if we say ui handling is identical for the time being with the exception that one shows an ayah number and one doesn't?)

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+      SuraPageItem(sura = quranDisplayData.getSuraNameString(appContext, it.page), page = it.page)
+    }
+  }
+
+  override fun onCreate() = Unit
+  override fun onDestroy() = Unit
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt:40]: remove

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  override fun getItemId(position: Int) = position.toLong()
+
+  override fun hasStableIds() = false
+
+  /*
+  *Similar to getView of Adapter where instead of View
+  *we return RemoteViews
+  *
+  */
+  override fun getViewAt(position: Int): RemoteViews {
```
Comment (ahmedre) [inline app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt:57]: is there no recycling that we can get with this?

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt (hunk 15 above)
Comment (ismail-s) [reply app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt:40]: I don't think I can, it's part of [the interface](https://developer.android.com/reference/android/widget/RemoteViewsService.RemoteViewsFactory).

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt (hunk 16 above)
Comment (ismail-s) [reply app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt:57]: I don't think so, [looking at the interface methods(https://developer.android.com/reference/android/widget/RemoteViewsService.RemoteViewsFactory#getViewAt(int)). It doesn't seem like the [`Adapter#getView`](https://developer.android.com/reference/android/widget/Adapter#getView(int,%20android.view.View,%20android.view.ViewGroup)) method.

**Comment by ismail-s** [comment]
@ahmedre For the search box, did anything happen if you click on the search icon itself? It works on my phone (though I'm not that happy with it, you need to know to click there, I tried at the time to make the keyboard open automatically, but I remember I got other issues which stopped me from going further. I may need to revisit that).

**Comment by ahmedre** [comment]
this is what i see when i tap search -

Screenshot_1595371977

File: app/src/main/res/layout/search_widget.xml [hunk 17]
```diff
@@ -0,0 +1,34 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
[the remark this answers is not in the document]
Comment (ismail-s) [reply app/src/main/res/layout/search_widget.xml:1]: I've implemented this in a different way, see ca9efb1. I got the idea from [Orgzly app](https://github.com/orgzly/orgzly-android), though they aren't dealing with a DialogFragment, so can do things slightly simpler in a way.

File: app/src/main/AndroidManifest.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (ismail-s) [reply app/src/main/AndroidManifest.xml]: I've updated/added tests in `BookmarkModelTest`.

File: app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (ismail-s) [reply app/src/main/java/com/quran/labs/androidquran/BookmarksWidget.kt]: I've removed this and it still seems to work fine.

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (ismail-s) [reply app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: Did some thinking around this and tried one or two different things, ended up doing [this](https://github.com/quran/quran_android/pull/1378/files/95b7ccb090a04009f9f884ed23419a8b8d8ada3a..8d0174be051ca80813a348c9af404d9e9ab40d7b).

Widgets don't seem to keep their apps running in the background (I think you essentially give a list of instructions to the launchers and they create the views that make up the widgets for you), so on app startup, at some point I think the app needs to subscribe `BookmarksWidget` up to any updates to bookmarks. I've tried to do this in the simplest/minimal way possible.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (ismail-s) [reply app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt]: You're right, it can be done, and I've done this in 8976ec0741a6ecdc2d43ab99f93d23f0c462c2fd and 3b7df4e520755ac14bb94071c8cc34443c34070d.

File: app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt (hunk 1 above)
Comment (ismail-s) [reply app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt]: I made this change.

File: app/src/main/java/com/quran/labs/androidquran/widget/BookmarksWidgetUpdater.kt [hunk 18]
```diff
@@ -0,0 +1,26 @@
+package com.quran.labs.androidquran.widget
```
Comment (ismail-s) [inline app/src/main/java/com/quran/labs/androidquran/widget/BookmarksWidgetUpdater.kt:1]: These files I moved into a `widget` subpackage. There's already a subpackage called `widgets`, maybe that subpackage should be renamed to `views` instead?

**Comment by ismail-s** [comment]
I haven't addressed all comments yet, just focussed on the harder ones. Search should be working better now. Test that is failing in [Travis CI build](https://travis-ci.com/github/quran/quran_android/builds/178155091), I'm not sure what's going on, probs something with the RxJava scheduling/waiting. I tried switching to trampoline executor to make tests more reliable, it all works fine on my machine, not sure about CI.

**Comment by ahmedre** [comment]
test failure looks like it's from this test:

```
015com.quran.labs.androidquran.model.bookmark.BookmarkModelTest > testSubscribeBookmarksWidgetIfWidgetGetsAdded FAILED
3016    Wanted but not invoked:
3017    bookmarksWidgetUpdater.updateBookmarksWidget();
3018    -> at com.quran.labs.androidquran.model.bookmark.BookmarkModelTest.testSubscribeBookmarksWidgetIfWidgetGetsAdded(BookmarkModelTest.java:85)
3019
3020    However, there were exactly 2 interactions with this mock:
3021    bookmarksWidgetUpdater.checkForAnyBookmarksWidgets();
3022    -> at com.quran.labs.androidquran.model.bookmark.BookmarkModel.subscribeBookmarksWidgetIfNecessary(BookmarkModel.java:52)
3023
3024    bookmarksWidgetUpdater.checkForAnyBookmarksWidgets();
3025    -> at com.quran.labs.androidquran.model.bookmark.BookmarkModel.subscribeBookmarksWidgetIfNecessary(BookmarkModel.java:52)
3026        at com.quran.labs.androidquran.model.bookmark.BookmarkModelTest.testSubscribeBookmarksWidgetIfWidgetGetsAdded(BookmarkModelTest.java:85)
```

File: app/src/main/java/com/quran/labs/androidquran/widget/BookmarksWidgetUpdater.kt (hunk 18 above)
Comment (ahmedre) [reply app/src/main/java/com/quran/labs/androidquran/widget/BookmarksWidgetUpdater.kt:1]: `view` singular, yes - but that can be done later outside of this PR

File: app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (ahmedre) [reply app/src/main/java/com/quran/labs/androidquran/database/BookmarksDBAdapter.java]: i think this is good but not sure this should live in `BookmarksModel` - seems like bookmarks model should only deal with bookmarks, and the widget updater is something that should (and does) subscribe to it so should not live in this class (but should live in its own class). though this way, you'd have to "start up" that class whenever the app starts if widgets are enabled as you mentioned.

**Comment by ahmedre** [comment]
masha'Allah looking good - jump is great and search now works (also works without the widget al7amdulillah). minor things:

- in the smallest size of the large widget, we need some padding around "no bookmarks currently set" - maybe a smaller font also etc?

Screenshot_1596413834

- not sure if we can fix this, but when searching from the widget, the cursor is to the left of the magnifying glass until typing starts.

Screenshot_1596413893

also added a minor comment about the `BookmarkModel` change.

barak Allah feekum and jazakumAllah khairan for this change! really appreciate it and masha'Allah it's looking good - in sha' Allah when ready let me know and i can do another pass of a review and we can get this merged soon in sha' Allah.

**Comment by ismail-s** [comment]
@ahmedre I took a look into the magnifying glass cursor position thing, I don't think we can easily do anything about it, it looks like just how the SearchView works (it puts the magnifying glass icon in the underlying TextView, not as a separate ImageView, which is why the cursor appears before it).

I think I've addressed other remaining comments (sorry for taking a long time to get round to it). I found [this article](https://medium.com/@peter.tackage/overriding-rxandroid-schedulers-in-rxjava-2-5561b3d14212) which gave me enough understanding to do 0721824 (random fact: assuming SHA hashes are fully random, then the chances that commit started with 7 numbers is [0.0001](https://www.google.com/search?hl=en&q=(10%2F36)%5E7) or 1 in 10000) to sort out the flaky unittest and get it to use `TestScheduler` instead.

File: app/src/main/res/layout/search.xml (hunk 5 above)
Comment (ozbek) [reply app/src/main/res/layout/search.xml:15]: Not directly related, but IMHO this should actually go to toolbar menu.

Comment (ismail-s) [reply app/src/main/res/layout/search.xml:15]: I agree (if only I'd thought of this before), I'm taking a look into this, I've got something mostly working, not the prettiest code at the moment though. I want to test more around it and see if I can improve the code first.

Comment (ismail-s) [reply app/src/main/res/layout/search.xml]: @ozbek I've pushed f72bc5b and f82d4618.

File: app/src/main/res/values/strings.xml [hunk 19]
```diff
@@ -410,4 +410,7 @@
   <!-- notification channel names [CHAR LIMIT=40] -->
   <string name="notification_channel_download">Quran Downloads</string>
   <string name="notification_channel_audio">Quran Recitation</string>
+
+  <!-- Widgets -->
+  <string name="no_bookmarks">No bookmarks currently set</string>
```
Comment (ozbek) [inline app/src/main/res/values/strings.xml:415]: Can you please merge this with `bookmarks_list_empty` string? Perhaps, simply by rename that one. If you want, you  may rephrase it also. E.g.:

```
<string name="no_bookmarks">No bookmarks yet</string>
```

**Comment by ozbek** [comment]
Out of curiosity, what's the reason not having a jump button on the widget with bookmarks?

File: app/src/main/res/values/strings.xml (hunk 19 above)
Comment (ismail-s) [reply app/src/main/res/values/strings.xml]: If it's fine with you, I've deleted `no_bookmarks` and am just using `bookmarks_list_empty` as-is. Wording seems good enough to me, and not changing the wording means no updating translations.

Also, I couldn't find any other code using `bookmarks_list_empty`, so it's good you've spotted this.

**Comment by ismail-s** [comment]
@ozbek Not much reason, at the time it would have taken a few minutes longer and made the widget a bit wider. I've given it a try, it looks ok, so have pushed commits to add it.

**Comment by ozbek** [comment]
Awesome, jazak Allah khair for your efforts!

**Comment by ahmedre** [comment]
awesome akhi, jazakumAllah khair - search looks a lot better now and everything is good - merging this
